juju/juju
# feat(macaroon): wire up macaroon root key service

feat(macaroon): wire up macaroon root key service

We recently implemented a DQLite backed service for macaroon root keys
which implements dbrootkeystore.Backing & dbrootkeystore.ContextBacking.

Use this service to replace the old Mongo-backed backing.

Create an internal macaroon package, transfering core/macaroons
internally, move the ExpirableStorage interface, constructor and
implementation here out of state.
    
We can now drop state entirely from a few places.

Remove the old Mongo implementation

## Checklist


## QA steps

```
./main.sh -v cmr
```

Cherry pick the following commits to hotfix a known bug with external users and build juju:
- https://github.com/juju/juju/pull/17752/commits/12437cf19617af9808d546e8fc2a950bd95636e3
- https://github.com/juju/juju/pull/17772/commits/1d32c0e144565c6d445d356c226914b818df5f1b

Copy your ec2 credential over to `./cred.yaml` such that:
```
cat ./cred.yaml
credentials:
  aws:
    ec2-robot-external:
      auth-type: access-key
      access-key: ...
      secret-key: ...
```

The run + observe:
```
$ juju bootstrap aws/eu-west-2 aws --config identity-url=https://api.jujucharms.com/identity --config allow-model-access=true
$ juju grant <lp user>@external superuser
$ juju change-user-password
$ juju logout
$ juju login --user <lp user>@external
Welcome, <lp user>@external. You are now logged into "lxd".

Current model set to "admin/controller".

$ juju whoami
Controller:  aws
Model:       admin/controller
User:        <lp user>@external

$ juju add-credential aws -f ./cred.yaml --controller aws --client
Using cloud "aws" from the controller to verify credentials.
Credential "ec2-robot-external" added locally for cloud "aws".

Controller credential "ec2-robot-external" for user "jack-shaw@external" for cloud "aws" on controller "aws" added.
For more information, see ‘juju show-credential aws ec2-robot-external’.

$ juju add-model m --credential ec2-robot-external
Uploading credential 'aws/jack-shaw@external/ec2-robot-external' to controller
Added 'm' model on aws/eu-west-2 with credential 'ec2-robot-external' for user 'jack-shaw'

$ juju deploy ubuntu
$ juju status
Model  Controller  Cloud/Region   Version      Timestamp
m      aws         aws/eu-west-2  4.0-beta4.1  14:46:11+01:00

App     Version  Status   Scale  Charm   Channel        Rev  Exposed  Message
ubuntu           waiting    0/1  ubuntu  latest/stable   24  no       waiting for machine

Unit      Workload  Agent       Machine  Public address  Ports  Message
ubuntu/0  waiting   allocating  0        18.134.209.114         waiting for machine

Machine  State    Address         Inst id              Base          AZ          Message
0        pending  18.134.209.114  i-0790d125c69f15f77  ubuntu@22.04  eu-west-2b  running
```

NOTE: There are still some bugs with external users. They will hopefully be resolved with https://github.com/juju/juju/pull/17752 + other work
